service/controller/v26/resource/cleanupendpointips/create.go

```diff
@@ -0,0 +1,169 @@
+package cleanupendpointips
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/errors/tenant"
+	"github.com/giantswarm/k8sclient"
+	"github.com/giantswarm/kvm-operator/service/controller/v26/key"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/tenantcluster"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	// At first we need to create a Kubernetes client for the reconciled tenant
+	// cluster.
+	var k8sClient kubernetes.Interface
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "creating Kubernetes client for tenant cluster")
+
+		i := key.ClusterID(customObject)
+		e := key.ClusterAPIEndpoint(customObject)
+
+		restConfig, err := r.tenantCluster.NewRestConfig(ctx, i, e)
+		if tenantcluster.IsTimeout(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not create Kubernetes client for tenant cluster")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "waiting for certificates timed out")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+		clientsConfig := k8sclient.ClientsConfig{
+			Logger:     r.logger,
+			RestConfig: restConfig,
+		}
+		k8sClients, err := k8sclient.NewClients(clientsConfig)
+		if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant cluster is not available")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		k8sClient = k8sClients.K8sClient()
+		r.logger.LogCtx(ctx, "level", "debug", "message", "created Kubernetes client for tenant cluster")
+	}
+
+	// We need to fetch the nodes being registered within the tenant cluster's
+	// Kubernetes API.
+	var nodes []corev1.Node
+	{
+		list, err := k8sClient.CoreV1().Nodes().List(metav1.ListOptions{})
+		if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant cluster is not available")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+		nodes = list.Items
+	}
+
+	// Fetch the list of pods running on the control plane. These pods serve VMs
+	// which in turn run the tenant cluster nodes.
+	var pods []corev1.Pod
+	{
+		n := key.ClusterID(customObject)
+		list, err := r.k8sClient.CoreV1().Pods(n).List(metav1.ListOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		pods = list.Items
+	}
+	// Clear dead endpoint IPs only when the cluster is not in transitioning state.
+	// The amount of nodes should be equal to amount of pods.
+	if len(nodes) == len(pods) {
```

Comment:
This is still weird and the explanations so far were a bit counterintuitive. One argument was we need to check for transitions but I miss the reason why. Then one argument was we need to run the cleanup especially during upgrades, which is a transition. So all this is confusing and I am not sure what is the right thing to do here now.

Reply:
Alright I will rewrite and change the comment